storagemarket/impl/provider_states.go

```diff
@@ -195,12 +202,24 @@ func (p *Provider) staged(ctx context.Context, deal MinerDeal) (func(*MinerDeal)
 
 func (p *Provider) sealing(ctx context.Context, deal MinerDeal) (func(*MinerDeal), error) {
 	// TODO: consider waiting for seal to happen
+	cb := func(err error) {
+		select {
+		case p.updated <- minerDealUpdate{
+			newState: storagemarket.DealComplete,
+			id:       deal.ProposalCid,
+			err:      err,
+		}:
+		case <-p.stop:
```

Comment:
I can't tell from looking at this PR where it is that you send a value to `p.stop` - but I do see that you've got at least one anonymous goroutine (elsewhere in the file) which uses receipt of a value on `p.stop` to signal exit. 

If it's the case that a single termination value on `p.stop` is sent - then I fear that you'll end up with orphaned goroutines (only one will receive the termination value).

Reply:
p.stop sends by being closed, which means it will read multiple times.

Reply:
Nice. I didn't know that about Golang channels.